File: dual_dither.m
if 1
    clear
    SYSTEM = 7;
else
    SYSTEM = 0;
end

N = 1e6;
Fs = 1e6;
Ts = 1/Fs;

% v = randn(1,N);

%% specifying S(omega) = G*G'; Fourier transform of R(tau)

switch SYSTEM
    case 1 % random filter alt. 1
        g = randn(1,5);
        G_tf = tf(g,1,Ts,'Variable','z^-1'); % phi realisation
        S_num = conv(g,fliplr(g));
        S_den = [0 0 0 0 1];
    case 2 % random filter alt. 2
        G = drss(7);
        G_tf = tf(G);
        S = G*G';
        S_tf = tf(S); S.Ts = Ts;
        S_num = S_tf.Numerator{:};
        S_den = S_tf.Denominator{:};
    case 3
        g = [1 -1];
        G_tf = tf(g,1,Ts,'Variable','z^-1');
        S_num = conv(g,fliplr(g));
        S_den = 1;
    case 4
        g = [1 1];
        G_tf = tf(g,1,Ts,'Variable','z^-1');
        S_num = conv(g,fliplr(g));
        S_den = 1;
    case 5
        g = ones(1,10)/10;
        G_tf = tf(g,1,Ts,'Variable','z^-1');
        S_num = conv(g,fliplr(g));
        S_den = 1;
    case 6
        Nh = 2047;
        n = (0:Nh-1)-(Nh-1)/2;
        fc = 0.5/2; % cutoff frequency = 0.5
        g_lp = 2*fc*sinc(2*fc*n).*kaiser(Nh,50).'; % windowed sinc function
        %figure; freqz(h_lp)
        
        g = -g_lp; % spectral inversion
        g((Nh+1)/2) = g((Nh+1)/2) + 1;
        %figure; freqz(g), return

        Gtf = tf(g,1,Ts,'Variable','z^-1');

        S_num = conv(g,fliplr(g));
        S_den = 1;
    case 7 %
        Nh = 63;
        g = fir1(Nh,[0.5 0.8]);
        G_tf = tf(g,1,Ts,'Variable','z^-1');

        S_num = conv(g,fliplr(g));
        S_den = 1;
        
        
        %Fc = 100e3;
        %Fn = Fs/2;
        %[b,a] = butter(4, [Fc, 2*Fc]/Fn);  % Butterworth filter
        %G = tf(b,a,1/Fs);
        %Gc = tf([1 0.9 0.25],[1 0.001 2.5]); % tf([1 0.01 0.1],[0.1 0.01 1]); % tf([0.1 0.01],[0.01 0.1]); % tf([1 0.9],[0.05 1]);
        %Gc = balreal(Gc);
        %G = c2d(Gc,Ts,'zoh');
        %S = G*G';
        
        %S_tf = tf(S);
        %S_num = S_tf.Numerator{:};
        %S_den = S_tf.Denominator{:};
        

    case 8 % Valid PSD-shaping examples (Phi_(w) >= 0 for all w, condition is met)
        PSD_Example_Data = readmatrix('Valid_PSD_Shaping_Examples.txt');
        rand_index = size(PSD_Example_Data);
        rand_index = randi(rand_index(1)/2);
        S_num = PSD_Example_Data(2*rand_index-1,:);
        S_den = PSD_Example_Data(2*rand_index,:);  
    case 9 % Invalid PSD-shaping examples (Phi_(w) >= 0 for all w, condition is violated)
        PSD_Example_Data = readmatrix('Invalid_PSD_Shaping_Examples.txt');
        rand_index = size(PSD_Example_Data);
        rand_index = randi(rand_index(1)/2);
        S_num = PSD_Example_Data(2*rand_index-1,:);
        S_den = PSD_Example_Data(2*rand_index,:); 
    otherwise

end

M = 1024; % no. of frequency samples

%% Non-linearities chain parameters

B = 7; % Number of quantization bits
V_max = 10; % Dynamic range - upper limit
V_min = -10; % Dynamic range - lower limit
FSR = V_max-V_min; % Full Scale Range
LSB = FSR/(2^B-1); % Least Sitgnificant Bit
Quant_L = V_min:LSB:V_max; % Quantized Output Levels (2^B)
T = Quant_L(1:end-1)+0.5*LSB; % Quantization Thresholds (Mid-rise, 2^B-1)
Var_v = 1; % Initial dist. variance
Var_Eq = LSB^2/12; % Quantization Error variance (Output stage - Uniform Quantizer)

pdf_initial = makedist('Normal','mu',0,'sigma',sqrt(Var_v)); % Initial distribution


Amp = 50*LSB; % Amp value for a desired Uniform PDF [LSB]
pdf_des = makedist('Uniform',-Amp,Amp); % Desired dither probablity deistribution (RPDF)
Var_y = pdf_des.var; 


%% frequency response of S(omega)

% M = 1024; % no. of frequency samples
w = linspace(-pi,pi,M); % sample whole circle
S_fr = freqz(S_num,S_den,w); % frequency response samples

% if SYSTEM==7
%     S_fr = freqresp(S,w,'rad/s');
%     S_fr = S_fr(:);
%     S_fr = circshift(S_fr,M/2);
% end

w = linspace(0,2*pi,M); % sample whole circle

%% determining the norm/variance and analytical S(omega)

% The supplied norm() function in MATLAB fails 
S_fr_2norm = sum(abs(S_fr).*mean(diff(w)))/(2*pi);
S_fr_ = (S_fr/S_fr_2norm)*Var_y; % scale response to a feasible variance at the output of the non-linear element [v+qUniform: (LSB^2/12)]

% var(y) = norm(G)^2, when y = G v, and v unity variance white noise


%% PSD S(w)

S_fr = abs(S_fr);
N_fir = 256;
[R_,R_win,R_beta] = fir_filter_ls(S_fr,N_fir);
S_fir_fr = S_num; %freqz(R_win,R_beta,w); % frequency response samples

% figure(45)
% plot(w/(2*pi), 10*log10(abs(S_fr)), w/(2*pi), 10*log10(abs(S_fir_fr)))

%% compute phi
R_win_ = 0.999*R_win/(norm(R_win,Inf)+sqrt(eps));

% phi = sin(pi/2*R_win_); % eq (16c) in Sondhi 1983 [Binary]
phi = 2*sin((pi/6)*R_win_); % eq (16a) in Sondhi 1983 [Uniform]

% plot(phi,R_win_,p,R)

Phi_fr = freqz(phi,1,w); % frequency response

switch 1
    case 1 % use FFT/IFFT to synth. H (same as Sondhi - 1983)
        phi_ = circshift(phi,128);
        % figure(50), stem(phi_)

        Phi = fft(phi_);

        % if(real(Phi)==abs(real(Phi)))
        %     Table_num = table(S_num); % Examples of well-behaved desired PSDs
        %     Table_denum = table(S_den); % Examples of well-behaved desired PSDs
        %     writetable(Table_num, 'Dither_Gen_HP_PSD_Shaping_Examples.txt','WriteMode','Append','WriteVariableNames',false); % Record randomly generated feasible PSD examples.
        %     writetable(Table_denum, 'Dither_Gen_HP_PSD_Shaping_Examples.txt','WriteMode','Append','WriteVariableNames',false);
        % else
        %     Table_num = table(S_num); % Examples of ill-behaved PSDs
        %     Table_denum = table(S_den); % Examples of ill-behaved PSDs
        %     writetable(Table_num, 'Dither_Gen_Invalid_PSD_Shaping_Examples.txt','WriteMode','Append','WriteVariableNames',false); % Record randomly generated feasible PSD examples.
        %     writetable(Table_denum, 'Dither_Gen_Invalid_PSD_Shaping_Examples.txt','WriteMode','Append','WriteVariableNames',false);
        % end

        % Th = w(real(Phi)<0) & w(S_fr_<=Var_Eq);
        
        Phi(real(Phi)<0) = 0; % (Filter Approx. Eq(18) in Sondhi - 1983) ***

        mus = sqrt(abs(Phi));

        figure(55)
        subplot(2,1,1)
        plot(real(fft(circshift(R_win_,128)))), hold on, plot(real(Phi)), plot(real(mus)), hold off
        str = '$PSD [dB/Hz]$';
        ylabel(str,'Interpreter','latex')
        legend('$Re[S]$','$Re[\Phi]$','$Re[\mu]$','interpreter','latex')
        subplot(2,1,2)
        plot(imag(fft(circshift(R_win_,128)))), hold on, plot(imag(Phi)), plot(imag(mus)), hold off
        str = '$\arg{PSD} [deg]$';
        ylabel(str,'Interpreter','latex')
        legend('$Im[S]$','$Im[\Phi]$','$Im[\mu]$','interpreter','latex')
        str = '$\tau$';
        xlabel(str,'Interpreter','latex')

        

        h = ifft(real(mus)); % real(mus) excludes real(Phi) < 0
        h = circshift(h,128);

        % figure(60), stem(h)
    case 2 % use LS on frequency response to synth. H
        
        mus = sqrt(real(Phi_fr));
        [h_alpha,h_alpha_win,h_beta] = fir_filter_ls(mus,1000);

        h = h_alpha_win;
end

Phi_fr_2norm = sum(abs(Phi_fr).*mean(diff(w)))/(2*pi); % should be 1

figure(100)
%plot(w/(2*pi), 10*log10(abs(R_fir_fr)), w/(2*pi), 10*log10(abs(phi_fir_fr)))
plot(w/(2*pi), 10*log10(abs(S_fr_)))
hold on
plot(w/(2*pi), 10*log10(Var_y*abs(Phi_fr)))
plot(w/(2*pi),10*log10(Var_y)*ones(numel(w/(2*pi)),1),'--','Color',[0.4660 0.6740 0.1880],'LineWidth',1.5)
plot(w/(2*pi),10*log10(Var_Eq)*ones(numel(w/(2*pi)),1),'-.','Color',[0.9290 0.6940 0.1250],'LineWidth',1.5)
hold off
grid on
legend('$S(\omega)$','$\Phi(\omega)$','NSD noise floor','SD noise floor','Location','best','interpreter','latex')

figure(200)
stem(R_win_)
hold on
stem(phi)
hold off
str = '$\tau$';
xlabel(str,'Interpreter','latex')
str = '$ACF$';
ylabel(str,'Interpreter','latex')
legend('$R(\tau)$','$\rho(\tau)$','interpreter','latex')

%% generate coloured input to non-lin.

v = random(pdf_initial, 1, N);

switch 1
    case 1 % use corrected filter
        x_ = filter(h,1,v);
    case 2 % use "original" filter
        x_ = filter(g,1,v);
end

v_hat = sqrt(var(v)/var(x_))*x_; % normalise (should not be neccessary)
x_ = pdf_des.icdf(pdf_initial.cdf(v_hat));

% % Plot Output Histogram & PSD
figure(5)
histogram(x_/LSB,'Normalization', 'pdf'); title('PDF-PSD Shaped Dither Ditribution');xlabel('Amplitude [LSB]');ylabel('PDF');
% 
figure(6)
WIN = kaiser(N/250,10);
plot(w(w < pi)/(2*pi), 10*log10(2*abs(S_fr_(w < pi))))
% trapz(w(w < pi)/(2*pi),2*abs(S_fr_(w < pi))) % Area under the PSD curve - Compare to Var_y (Should/ Must add up)
hold on
[Pdd,Fd] = pwelch(x_,WIN,[],[],[],'psd','onesided'); % Fy(10*log10(Pyy)>=10*log10(2*Var_Eq))
% trapz(Fy,Pyy) % Area under the PSD curve - Compare to Var_y (Should/ Must add up)
plot(Fd,10*log10(Pdd))
plot(Fd,10*log10(2*Var_y)*ones(numel(Fd),1),'--','Color',[0.4660 0.6740 0.1880],'LineWidth',1.5)
hold off
grid on
legend('Desired','Acheived','NSD power level','Location','best','interpreter','latex')
str = 'Normalized frequency';
xlabel(str,'Interpreter','latex')
str = '$PSD [dB/Hz]$';
ylabel(str,'Interpreter','latex')
str = 'Desired vs. Acheived single-sided output PSD $S(\omega)$';
title(str,'Interpreter','latex')

File: utils/dual_dither.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Synthesise dual specification dither.

@author: Ahmad Faza, Arnfinn Eielsen
@date: 10.04.2024
@license: BSD 3-Clause
"""

import numpy as np
from numpy import linalg
from scipy import signal
from scipy import special
import matplotlib.pyplot as plt

from utils.fir_filter_ls import fir_filter_ls


def psd_fr_2norm(S, w):
    norm = np.sum(np.abs(S)*np.mean(np.diff(w)))/(2*np.pi)
    return norm


def dual_dither(N=int(1e6), make_plots=False):
    # Specify S(omega) = G*G'; Fourier transform of R(tau)
    M = 1024  # no. of frequency samples
    w = np.linspace(0, 2*np.pi, M)  # sample whole circle
    Ihlf = np.argwhere(w < np.pi)  # indices to half-circle

    match 8:
        case 1:
            S_fr = np.tan(w[Ihlf]/2.15 + 0.1)
            S_fr = np.r_[S_fr, np.flipud(S_fr)]  # make symmetric
        case 2:
            n = np.arange(0,M)-M/2
            fc = 0.5/2  # cutoff frequency = 0.5
            g = 2*fc*np.sinc(2*fc*n)*np.kaiser(M, 1)  # windowed sinc function
            g = -g # spectral inversion
            g[int(M/2)] = g[int(M/2)] + 1
            _, g_fr = signal.freqz(g, 1, w)
            S_fr = abs(g_fr)**2
        case 3:
            #win = np.hanning(int(M/2))
            win = np.kaiser(M/2, 5)
            S_fr = np.r_[win, np.flipud(win)]  # make symmetric
        case 4:
            #win = np.hamming(M)
            win = np.kaiser(M, 5)
            #win = signal.windows.gaussian(M, 175)
            S_fr = win
        case 5:
            hu = np.linspace(0.2, 2, int(w.size/2))
            hd = np.linspace(2, 0.2, int(w.size/2))
            S_fr = np.r_[hu**2, hd**2]
        case 6:
            MM = 256 
            hs = 0.0125*np.ones(int(MM))
            hp = 3*np.ones(int(M - 2*MM))
            S_fr = np.r_[hs, hp, hs]
        case 7:
            b, a = signal.butter(1, 0.2, btype='high', analog=False)#, fs=Fs)
            signal.freqz(b, a, w)
            _, g_fr = signal.freqz(b, a, w)
            S_fr = abs(g_fr)**2
        case 8:
            d = 1e-2
            wn = np.pi/3
            ws = w - np.pi
            S_fr = wn**2/np.sqrt(4*d**2*ws**2*wn**2 + ws**4 - 2*ws**2*wn**2 + wn**4)
            Irhlf = np.argwhere(w > np.pi)  # indices to half-circle
            S_fr = np.r_[np.flipud(S_fr[Irhlf]), S_fr[Irhlf]]  # make symmetric
            #S_fr = np.r_[S_fr[Irhlf], np.flipud(S_fr[Irhlf])]  # make symmetric

    # Determining the norm/variance and analytical S(omega)
    S_fr_2norm = psd_fr_2norm(S_fr, w) 
    S_fr_ = (S_fr/S_fr_2norm)*(4/12);  # scale response to correct variance (uniform pdf)
    # recall var(y) = norm(G)^2, when y = G v, and v unity variance white noise

    if make_plots:
        fig, ax1 = plt.subplots()
        ax1.set_title('$S(\omega)$ prototype frequency response')
        ax1.plot(w.squeeze(), 10*np.log10(np.abs(S_fr.squeeze())))
        ax1.grid(True)

    # Synthesise FIR filter
    S_fr_pr = np.abs(S_fr).squeeze()  # hack to ensure positive real
    N_fir = 1024
    [R, R_win, R_beta] = fir_filter_ls(S_fr_pr, N_fir)

    wS, S_fir_fr = signal.freqz(R_win, R_beta, w)  # frequency response samples

    if make_plots:
        fig, ax2 = plt.subplots()
        ax2.set_title('$S(\omega)$ prototype vs. FIR approximation')
        ax2.plot(w.squeeze(), 10*np.log10(np.abs(S_fr.squeeze())), wS.squeeze(), 10*np.log10(np.abs(S_fir_fr)))
        ax2.grid(True)
    
    # Compute phi(omega)
    R_win_inf = np.max(R_win) + np.sqrt(np.finfo(float).eps)  # inf norm
    R_win_ = R_win/R_win_inf
    phi = 2*np.sin((np.pi/6)*R_win_)  # compensation filter coeffs.

    if make_plots:
        fig, ax3 = plt.subplots()
        ax3.set_title('$R$ filter vs. $\phi$ filter coeffs.')
        ax3.stem(np.real(phi), 'b')
        ax3.stem(np.real(R_win_), 'r')
        plt.show()

    wphi_tf, Phi_fr = signal.freqz(phi, 1, w)

    match 1:
        case 1:  # use FFT/IFFT to synth. H (same as Sondhi - 1983)
            # phi_ = circshift(phi,128)
            phi_ = np.roll(phi, int(phi.size/2))
            
            # plot coeffs.
            if make_plots:
                fig, ax4 = plt.subplots()
                ax4.set_title('$\phi$ circularly shifted filter coeffs.')
                ax4.stem(phi_)
                plt.show()

            Phi = np.fft.rfft(phi_)
            #Phi = np.fft.fft(phi_)
            
            Phi_Re = np.real(Phi)
            Inp = np.argwhere(Phi_Re < 0)
            Phi_Re[Inp] = 0

            mus = np.sqrt(np.real(Phi_Re))  # realisable filter FFT

            if make_plots:  # plot FFTs
                fig, (ax5, ax6) = plt.subplots(2)
                fig.suptitle('rfft real and imag components of $\Phi$ and $\mu$')
                ax5.plot(np.real(Phi))
                ax5.plot(np.real(mus))
                ax6.plot(np.imag(Phi))
                ax6.plot(np.imag(mus))
                plt.show()

            h = np.fft.irfft(np.real(mus)) #
            #h = np.fft.ifft(np.real(mus)) #
            h = np.roll(h, int(phi.size/2))

            if make_plots:
                fig, ax7 = plt.subplots()
                ax7.set_title('$h$ comp. filter coeffs.')
                ax7.stem(h)
                plt.show()
        case 2:  # use LS on frequency response to synth. H
            mus = np.sqrt(abs(Phi_fr))
            [h_alpha,h_alpha_win,h_beta] = fir_filter_ls(mus, 1000)
            h = h_alpha_win

    Phi_fr_2norm = psd_fr_2norm(Phi_fr, w)  # verification; should be 1

    if make_plots:
        fig, ax8= plt.subplots()
        ax8.set_title('$S(\omega)$ vs. $\Phi(\omega)$')
        ax8.plot(w/(2*np.pi), 10*np.log10(np.abs(S_fr)))
        ax8.plot(w/(2*np.pi), 10*np.log10(np.abs(Phi_fr)))
        plt.show()

    # generate coloured input to non-lin.
    v = np.random.normal(0, 1.0, N)  # normally distr. noise

    #match 1:
    #    case 1: # use comp. filter
    zi = signal.lfilter_zi(h, 1)
    z, _ = signal.lfilter(h, 1, v, zi=zi*v[0])
            #x_ = filter(h, 1, v)
    #    case 2: # use "original" filter, not impl.
    #        zi = signal.lfilter_zi(g, 1)
    #        z, _ = signal.lfilter(g, 1, v, zi=zi*v[0])
            #x_ = filter(g, 1, v)

    x = z/np.std(z) # normalise (should strictly not be neccessary)

    # non-lin. transform
    y = 2*(0.5*(1 - special.erf(-x/np.sqrt(2))) - 0.5)

    # %%
    if make_plots:
        hist_and_psd_cmp(w, Ihlf, S_fr_, y)

    return y

def hist_and_psd(y):
    # histogram to check PDF specification
    _ = plt.hist(y, bins='auto')  # arguments are passed to np.histogram
    plt.title("Histogram of non-lin. transform output")
    plt.show()
    
    # PSD to check spectral specifcation
    fy, Pyy = signal.welch(y, fs=1, nperseg=y.size/500)

    plt.plot(fy, 10*np.log10(Pyy))
    plt.xlabel('Frequency (Hz)')
    plt.ylabel('Power (V$^2$/Hz)')
    plt.show()


def hist_and_psd_cmp(w, Ihlf, S_fr_, y):
    # histogram to check PDF specification
    _ = plt.hist(y, bins='auto')  # arguments are passed to np.histogram
    plt.title("Histogram of non-lin. transform output")
    plt.show()
    
    # PSD to check spectral specifcation
    fy, Pyy = signal.welch(y, fs=1, nperseg=y.size/500)
    fn = w[Ihlf]/(2*np.pi)
    fn = fn.squeeze()

    PSS = 2*abs(S_fr_[Ihlf])  # analytical desired resp.
    PSS = PSS.squeeze()

    plt.plot(fy, 10*np.log10(Pyy))
    plt.plot(fn, 10*np.log10(PSS))
    plt.xlabel('Frequency (Hz)')
    plt.ylabel('Power (V$^2$/Hz)')
    plt.show()


def main():
    """
    Test the method.
    """
    
    y = dual_dither(N=int(1e6), make_plots=True)


if __name__ == "__main__":
    main()

# #Nh = 2047
# Nh = 1024
# #n = np.arange(0,Nh)-(Nh-1)/2
# n = np.arange(0,Nh)-Nh/2

# fc = 0.5/2  # cutoff frequency = 0.5
# g_lp = 2*fc*np.sinc(2*fc*n)*np.kaiser(Nh, 1)  # windowed sinc function

# w, g_lp_fr = signal.freqz(g_lp, 1)

# g = -g_lp # spectral inversion
# g[int(Nh/2)] = g[int(Nh/2)] + 1

# w, g_fr = signal.freqz(g, 1)

# fig, ax = plt.subplots()
# ax.plot(w, 20*np.log10(np.abs(g_lp_fr)))
# ax.grid(True)

# fig, ax = plt.subplots()
# ax.plot(w, 20*np.log10(np.abs(g_fr)))
# ax.grid(True)

# S_num = np.convolve(g, np.flipud(g))
# S_den = 1


# w, S_fr = signal.freqz(S_num, 1)

# fig, ax = plt.subplots()
# ax.plot(w, np.log10(np.abs(S_fr)))
# ax.plot(w, np.log10(np.abs(g_fr**2)))
# ax.grid(True)
# %%
